### PR TITLE
User traits editor UI

### PR DESCRIPTION
**Pull request by flyinghermit**
Add, edit or delete user traits from Web UI.

changelog: Added ability to edit user traits from Web UI.

### CODE REVIEW

File: lib/web/users.go [hunk 1]
```diff
@@ -169,7 +169,13 @@ func updateUser(r *http.Request, m userAPIGetter) (*ui.User, error) {
...
 	user.SetRoles(req.Roles)
 
-	updateUserTraits(req, user)
+	// we do not want situation where one trait set (allTraits)
+	// overrides another (traits).
+	if len(req.AllTraits) > 0 {
+		user.SetTraits(req.AllTraits)
+	} else {
+		updateUserTraits(req, user)
+	}
```
Comment (flyinghermit) [inline lib/web/users.go:178]: This PR makes use of `AllTraits` field to manage user Traits. 

I am guessing we would ideally want to either utilize `AllTraits` or `Traits` (which just selectively updates specific traits.  But tentatively tagging @marcoandredinis as Github shows you added user traits editing functionality in user API. Did I miss anything here?

**Comment by flyinghermit** [comment]
How the editor UI looks: 
Screenshot 2024-06-10 at 2 17 31 PM

File: lib/web/users.go [hunk 2]
```diff
@@ -169,7 +169,13 @@ func updateUser(r *http.Request, m userAPIGetter) (*ui.User, error) {
 
 	user.SetRoles(req.Roles)
 
-	updateUserTraits(req, user)
+	// we do not want situation where one trait set (allTraits)
+	// overrides another (traits).
```
Comment (r0mant) [inline lib/web/users.go:173]: I think the comment is slightly confusing, I would rephrase to indicate that in some cases clients of this API just send builtin traits to set and some send the entire traits set to set.

File: lib/web/users.go (hunk 1 above)
Comment (r0mant) [inline lib/web/users.go:178]: Is this API covered by unit tests? Can we cover this logic as well?

File: lib/web/users.go [hunk 3]
```diff
@@ -305,9 +311,10 @@ type userTraits struct {
 // - if the value is an empty array we remove every element from the trait
 // - otherwise, we replace the list for that trait
 type saveUserRequest struct {
-	Name   string     `json:"name"`
-	Roles  []string   `json:"roles"`
-	Traits userTraits `json:"traits"`
+	Name      string              `json:"name"`
+	Roles     []string            `json:"roles"`
+	Traits    userTraits          `json:"traits"`
+	AllTraits map[string][]string `json:"allTraits"`
```
Comment (r0mant) [inline lib/web/users.go:317]: Nit: What do you think about naming this `OverrideTraits` or `SetTraits`? To make it more clear that if it's provided, it will set all user traits?

Also, can you add godocs on these fields.

File: lib/web/users.go [hunk 4]
```diff
@@ -305,9 +311,10 @@ type userTraits struct {
...
 }
 
 func (r *saveUserRequest) checkAndSetDefaults() error {
```
Comment (r0mant) [inline lib/web/users.go:320]: `Traits` and `AllTraits` are mutually exclusive it seems like, we should update checkAndSetDefaults accordingly.

**Review by r0mant — COMMENTED** [review]
Also @flyinghermit what's the behavior for SSO users? The "edit" dialog is not available to them so they can't edit the traits, correct?

File: lib/web/users.go (hunk 1 above)
Comment (marcoandredinis) [reply lib/web/users.go:178]: From what I can see, Traits contains only the default used traits (logins, aws arn roles, windows logins, db users, db names, ...).
While AllTraits contains everything (I think it was introduced as part of Access Lists).

I don't have the Access List context, but given that Lisa and Roman are already your reviewers, I think you are good.

Comment (flyinghermit) [reply lib/web/users.go]: tests added in https://github.com/gravitational/teleport/pull/42620/commits/0d6b589bcda5f4c5525ffba6964a281b30d7025b#diff-182f90cb2b8f651ffc752d871baae2b6de277eec5a40fbb4cbbc28c8648b4676

File: lib/web/users.go (hunk 4 above)
Comment (flyinghermit) [reply lib/web/users.go:342]: updated checkAndSetDefaults in https://github.com/gravitational/teleport/pull/42620/commits/0d6b589bcda5f4c5525ffba6964a281b30d7025b

File: lib/web/users.go (hunk 3 above)
Comment (flyinghermit) [reply lib/web/users.go:339]: tl;dr: I updated `Traits` to `TraitsPreset` and added Godoc for clarity. Json tags are as is since they are also used in other parts of frontend. Lmk if you still find it confusing. 

The names `Traits` and `AllTraits` predates this PR and it isn't super straightforward to update these. I took a second tour today but there are other parts in frontend codebase (I found one discovery flow and not 100% on access list) that make use of these fields.

**Comment by flyinghermit** [comment]
> Also what's the behavior for SSO users? The "edit" dialog is not available to them so they can't edit the traits, correct?

Correct. The "options" menu, which wraps the edit button is not shown for the SSO users.

File: lib/web/users.go [hunk 5]
```diff
@@ -104,7 +104,16 @@ func createUser(r *http.Request, m userAPIGetter, createdBy string) (*ui.User, e
 
 	user.SetRoles(req.Roles)
 
-	updateUserTraits(req, user)
+	// checkAndSetDefaults makes sure either TraitsPreset
+	// or AllTraits field to be populated. Since empty
+	// AllTraits is also used to delete all user traits,
+	// we explicitly check if TraitsPreset is empty so
+	// to prevent traits deletion.
+	if req.TraitsPreset == (traitsPreset{}) {
```
Comment (r0mant) [inline lib/web/users.go:112]: That's a strange check, can you make it a pointer and then check against nil? And/or have an isEmpty method on the struct that returns true/false?

File: lib/web/users.go [hunk 6]
```diff
@@ -169,7 +178,16 @@ func updateUser(r *http.Request, m userAPIGetter) (*ui.User, error) {
...
```
Comment (r0mant) [inline lib/web/users.go:186]: Same comment as above?

File: lib/web/users.go [hunk 7]
```diff
@@ -317,5 +346,8 @@ func (r *saveUserRequest) checkAndSetDefaults() error {
 	if len(r.Roles) == 0 {
 		return trace.BadParameter("missing roles")
 	}
+	if len(r.AllTraits) != 0 && r.TraitsPreset != (traitsPreset{}) {
```
Comment (r0mant) [inline lib/web/users.go:349]: Same comment as above regarding the check.

**Review by r0mant — COMMENTED** [review]
Mostly lgtm but can we fix the strange traits preset check?

File: lib/web/users.go (hunk 5 above)
Comment (flyinghermit) [reply lib/web/users.go]: :D
Made the field pointer so now it can be compared with `nil` https://github.com/gravitational/teleport/pull/42620/commits/1a39fd7dee18702b11d3049dc6f72d303496be99

**Comment by flyinghermit** [comment]
Friendly ping

**Review by r0mant — APPROVED** [review]

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    setConfiguredTraits(newTraits);
+  }, [allTraits]);
+
+  function handleInputChange(i: InputOption) {
+    let newTraits = [...configuredTraits];
+    if (i.labelField === 'traitValues') {
+      let traitValue: Option[] = i.option as Option[];
+      newTraits[i.index] = {
+        ...newTraits[i.index],
+        [i.labelField]: [...traitValue],
```
[the remark this answers is not in the document]
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:72]: good catch, thanks.

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+}: TraitEditorProps) {
+  useEffect(() => {
+    const newTraits = traitsToTraitsOption(allTraits);
+    setConfiguredTraits(newTraits);
+  }, [allTraits]);
+
+  function handleInputChange(i: InputOption) {
+    let newTraits = [...configuredTraits];
+    if (i.labelField === 'traitValues') {
+      let traitValue: Option[] = i.option as Option[];
```
[the remark this answers is not in the document]
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:69]: This is needed because `InputOption.option` accepts both `Option | Option[];`. The alternative would be to define seperate type for input options
```
type InputOption = {
  labelField: 'traitKey';
  option: Option;
  index: number;
};

type InputOptionArray = {
  labelField: 'traitValues';
  option: Option[];
  index: number;
};
```

and do `function handleInputChange(i: InputOption | InputOptionArray) {`. But that still needs to specify type on handlechange 
```
onChange={e => {
    handleInputChange({
    option: e as Option[],
```
Not sure which one it better typed though! Any suggestions?

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+  configuredTraits,
+  setConfiguredTraits,
+}: TraitEditorProps) {
+  useEffect(() => {
+    const newTraits = traitsToTraitsOption(allTraits);
+    setConfiguredTraits(newTraits);
+  }, [allTraits]);
+
+  function handleInputChange(i: InputOption) {
+    let newTraits = [...configuredTraits];
```
[the remark this answers is not in the document]
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:67]: I have a mixed feeling on using `const` in this case. Cause, `const` semantically means the value wont change. But here we change `newTraits` value and that is only possible because its an array. So using `let` conveys that its value changes below.

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (kimlisa) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:69]: i would do exactly as you proposed, i think we are lacking typing on `FieldSelectCreatable` to specify like a generic type for options

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (kimlisa) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:58]: i can see how that can be confusing 😅 , but i think this [article](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) may help you out

tldr: 

> The const declaration creates an immutable reference to a value. It does not mean the value it holds is immutable — just that the variable identifier cannot be reassigned.

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.test.tsx [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        <TraitsEditor
+          allTraits={singleTrait}
+          attempt={{ status: '' }}
+          configuredTraits={traitsToTraitsOption(singleTrait)}
+          setConfiguredTraits={setConfiguredTraits}
+        />
+      </Validation>
+    );
+    fireEvent.click(screen.getByTitle('Remove Trait'));
+    expect(setConfiguredTraits).toHaveBeenCalledWith([]);
```
[the remark this answers is not in the document]
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.test.tsx]: This test block is not needed now as the  test case in first render is now no longer required since the `traitsToTraitsOption` is moved to  `useDialog.ts` now in https://github.com/gravitational/teleport/pull/42620/commits/314783e29151dc5c6c3035f684928298f6bf7000#diff-3e716db9176ae7ea0cef1294cddedf5c09b3d6f68b10e27ce6ecf676b49e9fcd.

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+    return { valid: true };
+  };
+
+export function traitsToTraitsOption(allTraits: AllUserTraits): TraitsOption[] {
+  let newTrait = [];
+  for (let trait in allTraits) {
+    if (!allTraits[trait][0]) {
+      continue;
+    }
+    if (allTraits[trait].length > 0) {
```
[the remark this answers is not in the document]
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx]: > also is it valid to add a trait with a empty name?

It shouldn't be a valid case but I know the `tctl` does not check this.

The `if (!allTraits[trait][0]) {` block exists because when user is created using tctl, some trait gets default `null` value and some, such as `host_user_gid` gets `""`. In the latter case, the value is essentially not a user configured value so I think it we don't need show them in traits editor as a "pre-configured" traits. 

Screenshot 2024-06-13 at 12 25 02 PM


Pinging @r0mant if there is any reason to allow traits with empty value.

[the remark this answers is not in the document]
Comment (r0mant) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx]: Can't think of any.

File: web/packages/teleport/src/Users/UserAddEdit/UserAddEdit.tsx [hunk 13]
```diff
@@ -80,37 +84,46 @@ export function UserAddEdit(props: ReturnType<typeof useDialog>) {
...
-              value={selectedRoles}
-              onChange={values => onChangeRoles(values as Option[])}
-              noOptionsMessage={() => 'No roles found'}
-              loadOptions={async input => {
-                const roles = await fetchRoles(input);
-                return roles.map(r => ({ value: r, label: r }));
-              }}
-              elevated={true}
-            />
+            <Box maxWidth={690}>
```
Comment (kimlisa) [inline web/packages/teleport/src/Users/UserAddEdit/UserAddEdit.tsx:91]: is this maxWidth required? (we already set a maxWidth on the parent)

File: web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx [hunk 14]
```diff
@@ -32,6 +34,9 @@ export default function useUserDialog(props: Props) {
       label: r,
     }))
   );
+  const [configuredTraits, setConfiguredTraits] = useState<TraitsOption[]>(
+    traitsToTraitsOption(props.user.allTraits)
+  );
```
Comment (kimlisa) [inline web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:39]: otherwise, `traitsToTraits` func will needlessly run every re-render
```suggestion
  const [configuredTraits, setConfiguredTraits] = useState<TraitsOption[]>(() =>
    traitsToTraitsOption(props.user.allTraits)
  );
```

File: web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx [hunk 15]
```diff
@@ -42,9 +47,16 @@ export default function useUserDialog(props: Props) {
   }
 
   function onSave() {
+    let traitsToSave = {};
```
Comment (kimlisa) [inline web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:50]:
```suggestion
    const traitsToSave = {};
```

File: web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx [hunk 16]
```diff
@@ -75,11 +87,14 @@ export default function useUserDialog(props: Props) {
     onChangeName,
     onChangeRoles,
     fetchRoles: props.fetchRoles,
+    setConfiguredTraits,
     isNew: props.isNew,
+    allTraits: props.user.allTraits,
```
Comment (kimlisa) [inline web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:92]: does this need to exported?

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+ * @param setConfiguredTraits sets user traits in current editor.
+ */
+export function TraitsEditor({
+  attempt,
+  configuredTraits,
+  setConfiguredTraits,
+}: TraitEditorProps) {
+  function handleInputChange(i: InputOption | InputOptionArray) {
+    const newTraits = [...configuredTraits];
+    if (i.labelField === 'traitValues') {
```
Comment (kimlisa) [inline web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:59]: we should not allow adding empty space b/c i can do this (and save it)

we can display it if it came like that, but the web UI can prevent adding further ones

image

**Review by kimlisa — APPROVED** [review]
after some minor comments, lgtm

File: web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx [hunk 18]
```diff
@@ -91,3 +106,22 @@ export type Props = {
   onCreate(user: User): Promise<any>;
   onUpdate(user: User): Promise<any>;
 };
+
+export function traitsToTraitsOption(allTraits: AllUserTraits): TraitsOption[] {
+  const newTrait = [];
+  for (let trait in allTraits) {
+    if (!allTraits[trait][0]) {
```
Comment (kimlisa) [inline web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:113]: actually, can we add a test for this function? wouldn't this throw if `allTraits[trait] = null`?
b/c i went to my staging cluster and saw this in my traits:
image

Comment (kimlisa) [reply web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:113]: for this condition `!allTraits[trait][0]` can this be possible?

```js
logins: ["", "root"]
```

if valid, this trait will get ignored?

File: web/packages/teleport/src/Users/UserAddEdit/UserAddEdit.tsx (hunk 13 above)
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/UserAddEdit.tsx:91]: This is required for inner content (added in https://github.com/gravitational/teleport/pull/42620/commits/7d4ecd8ed7807ce61c04b1ea64dafbce77eb2903). Without this, the inner content shrinks when vertical scrollbar appears, causing flickering.

File: web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx (hunk 17 above)
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:59]: I couldn't guess why would user save an empty trait value :D but I can surely avoid handing them footgun - https://github.com/gravitational/teleport/pull/42620/commits/c3b97772077a0dd6aedf919cd2e77ff1a33e5508#diff-72c47d20adf27d7885795d45757b748a86f288c7e018d043f6d0c9cac1f798ca

File: web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx (hunk 18 above)
Comment (flyinghermit) [reply web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx]: Tests added in https://github.com/gravitational/teleport/pull/42620/commits/1d38fb380b1465b91a7b0a724628a2df3f2e63ea

- `logins: ["", "root"]` will be shown as is. But `logins: [""]` will be filtered out. I am still sticking with this choice as in the backend, in its current state, that's a still default value not set by the user. 
- that didn't throw error cause `makeTraits` checks and sets empty array. Though, just to be on safe side, https://github.com/gravitational/teleport/pull/42620/commits/1e80fcff976f4a4647a2eb6f1e1c8eabfd843d43 now avoids null value explicitly.

File: web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx [hunk 19]
```diff
@@ -91,3 +105,27 @@ export type Props = {
...
+  const newTrait = [];
+  for (let trait in allTraits) {
+    if (!allTraits[trait]) {
+      continue;
+    }
+    if (!allTraits[trait][0]) {
+      if (allTraits[trait].length === 1) {
+        continue;
+      }
+    }
```
Comment (kimlisa) [inline web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:119]:
```suggestion
    if (allTraits[trait].length === 1 && !allTraits[trait][0]) {
        continue;
    }
```

**Review by kimlisa — APPROVED** [review]
LGTM, thanks for all the changes 👍
